src/coreclr/vm/unsafeaccessors.cpp

```diff
@@ -75,16 +93,249 @@ namespace
 
         UnsafeAccessorKind Kind;
         MethodDesc* Declaration;
-        MetaSig DeclarationSig;
+        SigPointer DeclarationSig;  // This is the official declaration signature. It may be modified
+                                    // to include the UnsafeAccessorTypeAttribute types.
+        MetaSig DeclarationMetaSig;
+        NewArrayHolder<ParamDetails> TranslatedParams; // Redefined types from UnsafeAccessorTypeAttribute usage.
+                                                       // Return type is at 0 index. Function arguments are 1 to N.
         SigPointer TargetTypeSig;
         TypeHandle TargetType;
         bool IsTargetStatic;
         MethodDesc* TargetMethod;
         FieldDesc* TargetField;
     };
 
+    void AppendTypeToSignature(
+        SigBuilder& sig,
+        TypeHandle th,
+        Instantiation inst)
+    {
+        STANDARD_VM_CONTRACT;
+        _ASSERTE(!th.IsNull());
+
+        //
+        // Building the signature follows details defined in ECMA-335 - II.23.2.12
+        //
+
+        // If we have any generic variables, mark as ELEMENT_TYPE_GENERICINST.
+        BOOL hasGenericVariables = inst.ContainsGenericVariables();
+        if (hasGenericVariables)
+            sig.AppendElementType(ELEMENT_TYPE_GENERICINST);
+
+        // Append the new type to the signature.
+        sig.AppendElementType(ELEMENT_TYPE_INTERNAL);
+        sig.AppendPointer(th.AsPtr());
+
+        // Append the instantiation types to the signature.
+        if (hasGenericVariables)
+        {
+            sig.AppendData(inst.GetNumArgs());
+            for (DWORD i = 0; i < inst.GetNumArgs(); i++)
+            {
+                if (!inst[i].IsGenericVariable())
```

Comment:
It does sound right to skip the generic variables. They should be encoded as generic variables.

Reply:
I'm not skipping generic variables. I'm only skipping if they aren't generic variables.

Reply:
Still, I would expect `ELEMENT_TYPE_GENERICINST` to be always followed by the type definition and N generic arguments, irrespective of whether those arguments are generic variables or not.

Reply:
Gotcha. Will update accordingly.